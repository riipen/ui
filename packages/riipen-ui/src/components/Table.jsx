import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";
import withForwardedRef from "../utils/withForwardedRef";

class Table extends React.Component {
  static propTypes = {
    /**
     * What background to render for the table.
     * Defaults to grey100.
     */
    backgroundColor: PropTypes.oneOf([
      "transparent",
      "grey50",
      "grey100",
      "grey200",
      "grey300",
      "grey400"
    ]),

    /**
     * Children to render in the table element.
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element.
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * A reference to the table DOM element.
     */
    forwardedRef: PropTypes.element,

    /**
     * Whether to give all columns in the table equal width.
     */
    layout: PropTypes.oneOf(["auto", "fixed"])
  };

  static defaultProps = {
    backgroundColor: "grey100",
    layout: "auto"
  };

  static contextType = ThemeContext;

  render() {
    const {
      children,
      classes,
      layout,
      backgroundColor,
      forwardedRef
    } = this.props;

    const theme = this.context;

    const tableBackground = {
      transparent: "transparent",
      grey50: theme.palette.grey[50],
      grey100: theme.palette.grey[100],
      grey200: theme.palette.grey[200],
      grey300: theme.palette.grey[300],
      grey400: theme.palette.grey[400]
    };

    return (
      <React.Fragment>
        <table ref={forwardedRef} className={clsx(classes)} border="0">
          {children}
        </table>
        <style jsx>{`
          table {
            background-color: ${tableBackground[backgroundColor]};
            border-radius: 4px;
            border-collapse: collapse;
            table-layout: ${layout};
            width: 100%;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withForwardedRef(withClasses(Table));
