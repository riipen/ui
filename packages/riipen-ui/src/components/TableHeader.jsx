import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import withClasses from "../utils/withClasses";

class TableHeader extends React.Component {
  static propTypes = {
    /**
     * Children to render in the thead element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string)
  };

  render() {
    const { children, classes, ...other } = this.props;
    return (
      <thead className={clsx(classes)} {...other}>
        {children}
      </thead>
    );
  }
}

export default withClasses(TableHeader);
