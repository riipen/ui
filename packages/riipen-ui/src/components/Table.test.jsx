import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import defaultTheme from "../themes/default";

import Table from "./Table";
import TableHeader from "./TableHeader";

describe("<Table>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Table />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Table />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<Table />);

      const component = wrapper.find("Table");
      expect(component.props().backgroundColor).toEqual("grey50");
      expect(component.props().layout).toEqual("auto");
    });
  });

  describe("backgroundColor prop", () => {
    it("sets custom valid backgroundColor with mapped hex color", () => {
      const color = "grey400";

      const wrapper = mount(<Table backgroundColor={color} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(
        `${defaultTheme.palette.grey[400]}`
      );
    });

    it("sets transparent color given transparent backgroundColor", () => {
      const color = "transparent";

      const wrapper = mount(<Table backgroundColor={color} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(color);
    });

    it("gives an error given an invalid backgroundColor", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const color = "white";

      mount(<Table backgroundColor={color} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("children prop", () => {
    it("renders all non null children", () => {
      const child = <TableHeader />;

      const wrapper = mount(
        <Table>
          {null}
          {child}
        </Table>
      );

      expect(wrapper.find("Table").contains(child)).toEqual(true);
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the root node", () => {
      const classes = ["classOne"];

      const wrapper = mount(<Table classes={classes} />);

      expect(
        wrapper
          .find("Table")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-table"];

      const wrapper = mount(<Table />);

      expect(
        wrapper
          .find("Table")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("layout prop", () => {
    it("applies valid custom layout", () => {
      const layout = "fixed";

      const wrapper = mount(<Table layout={layout} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(layout);
    });

    it("gives an error given an invalid layout", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const layout = "left";

      mount(<Table layout={layout} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });
});
