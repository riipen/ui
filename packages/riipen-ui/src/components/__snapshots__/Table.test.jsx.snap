// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table> renders correct snapshot 1`] = `
<ForwardRef(withForwardedRef(withClasses(Table)))>
  <withClasses(Table)
    forwardedRef={null}
  >
    <Table
      backgroundColor="grey50"
      classes={
        Array [
          "riipen",
          "riipen-table",
        ]
      }
      forwardedRef={null}
      layout="auto"
    >
      <table
        border="0"
        className="jsx-1059775881 riipen riipen-table"
      />
      <JSXStyle
        dynamic={
          Array [
            "#fafafa",
            "auto",
          ]
        }
        id="4123435624"
      />
    </Table>
  </withClasses(Table)>
</ForwardRef(withForwardedRef(withClasses(Table)))>
`;
